### PR TITLE
Initial groovy library support

### PR DESCRIPTION
**Pull request by grumpyjames**
Summary:
First cut of groovy_library support. Only very simple to begin with:
no cross compilation, no maven or source jar flavours.

Test-plan:

CI, but with extras: two new tests only provide coverage when
GROOVY_HOME is set; we recommend this be added to CI.

### CODE REVIEW

**Comment by grumpyjames** [comment]
One area of potential controversy is how we locate the groovy compiler, and the fact that it implicitly picks up its java from the environment's JAVA_HOME. We intend to revisit this as we add cross compilation, but if you feel it ought to go in with this change (or have a different view), let us know!

File: src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
+
+import java.nio.file.Path;
+
+public class GroovyLibraryDescription implements
+    Description<GroovyLibraryDescription.Arg>, Flavored {
+
+  public static final BuildRuleType TYPE = BuildRuleType.of("groovy_library");
+  public static final ImmutableSet<Flavor> SUPPORTED_FLAVORS = ImmutableSet.of();
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java]: Don't add the code for supporting flavours since it doesn't have any flavours yet, add it when you add the first flavour.

File: src/com/facebook/buck/jvm/groovy/GroovycStep.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    ProcessBuilder processBuilder = new ProcessBuilder(command.build());
+
+    // Set environment to client environment and add additional information.
+    Map<String, String> env = processBuilder.environment();
+    env.clear();
+    env.putAll(context.getEnvironment());
+    env.put("BUCK_INVOKING_RULE", invokingRule.toString());
+    env.put("BUCK_TARGET", invokingRule.toString());
+    env.put("BUCK_DIRECTORY_ROOT", filesystem.getRootPath().toAbsolutePath().toString());
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovycStep.java]: These three environment variables aren't necessary, you can remove them.

File: src/com/facebook/buck/jvm/groovy/GroovycStep.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    this.filesystem = filesystem;
+  }
+
+  @Override
+  public int execute(ExecutionContext context) throws IOException, InterruptedException {
+    ImmutableList.Builder<String> command = createCommandList();
+
+    ProcessBuilder processBuilder = new ProcessBuilder(command.build());
+
+    // Set environment to client environment and add additional information.
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovycStep.java]: Remove this comment as well.

File: src/com/facebook/buck/jvm/groovy/GroovycStep.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    return exitCode;
+  }
+
+  private ImmutableList.Builder<String> createCommandList() {
+    ImmutableList.Builder<String> command = ImmutableList.builder();
+    command.addAll(groovyc.getCommandPrefix(resolver))
+        .add("-cp")
+        .add(Joiner.on(":").join(Iterables.transform(
+            declaredClasspathEntries,
+            new Function<Path, String>() {
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovycStep.java]: `Functions.toStringFunction()`

File: src/com/facebook/buck/jvm/groovy/GroovycStep.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+              public String apply(Path input) {
+                return input.toString();
+              }
+            })))
+        .add("-d")
+        .add(outputDirectory.toString())
+        .addAll(extraArguments.or(ImmutableList.<String>of()))
+        .addAll(Iterables.transform(
+            sourceFilePaths,
+            new Function<Path, String>() {
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovycStep.java]: `Functions.toStringFunction()`

File: src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            new BuildTargetSourcePath(defaultJavaLibrary.getBuildTarget())));
+
+    return defaultJavaLibrary;
+  }
+
+  @SuppressFieldNotInitialized
+  public static class Arg {
+    public Optional<ImmutableSortedSet<SourcePath>> srcs;
+    public Optional<ImmutableSortedSet<SourcePath>> resources;
+    public Optional<ImmutableList<String>> extraArguments;
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java]: You should use a `MacroHandler` to support macros in here.

You can do that in a separate PR.

**Comment by Coneko** [comment]
Looks good, just a couple of notes.

File: src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java (hunk 6 above)
Comment (grumpyjames) [inline src/com/facebook/buck/jvm/groovy/GroovyLibraryDescription.java]: Will do.

**Comment by Coneko** [comment]
Cool! Thanks!

@sdwilsh Can you import this? Our bot seems to have some issues lately.

**Comment by sdwilsh** [comment]
@facebook-github-bot import

**Comment by sdwilsh** [comment]
> CI, but with extras: two new tests only provide coverage when
GROOVY_HOME is set; we recommend this be added to CI.

I strongly suggest you do this, as you are more likely to get it right than any of us :)

The travis config isn't too hard to modify.

**Comment by sdwilsh** [comment]
Any chance you'll be writing docs for this so others can use your work?  We generally don't talk about new features in our release notes (which I'm writing now) if it isn't documented.

**Comment by grumpyjames** [comment]
We will definitely add documentation, but we'd prefer not to _yet_. The next step (to introduce cross compilation with javac) makes a breaking change to how groovy_library is configured. We'd rather get that change in and document afterward rather than expose things now only to break them in the next release.

We can try to backport the config change to the initial version and document earlier if you want - let us know if that's your preferred route and we'll get to it ASAP.

**Comment by sdwilsh** [comment]
Nope, that's fine.  I just wanted to make sure it was on your radar.  I 
want you folks to get credit for your work in our release notes :)
